HIP/hip-1037.md

```diff
@@ -0,0 +1,95 @@
+usable or useful specification. We should add, in comments compatible with
+protoc-gen-doc, specification text that clearly states both requirement and
+expectation for every message, file, and field. This documentation should be
+automatically transformed to markdown documentation that is published to the
+Hedera API site. In the ideal case this specification will be sufficient for
+any qualified team to implement a completely independent consensus node that
+is fully interoperable with the current consensus node software.
+
+## Motivation
+The existing protocol buffer API documentation is brief, entirely descriptive,
+and often inaccurate. This produces two negative consequences. First, it is not
+possible to produce a compatible independent implementation of the Hedera
+consensus node software without extensive reference to the existing source
+code. Second, developers are often unclear how best to interact with the
+Hedera API, and independent SDK developers, in particular, may struggle to
+correctly implement these API calls.
+As part of expanding the Hedera ecosystem we MUST reduce the difficulty and
+hurdles for independent developers when building systems that interact with
+the Hedera network. Clear and accurate API specifications are one element
+required to accomplish this goal.
+
+## Rationale
+This HIP improves the developer experience and enhances decentralization by
+making it easier to design distributed applications or alternative
+implementations of core network systems.
+The use of a markdown format that is automatically generated from the
+functional protocol buffer definition files encourages more frequent updates
+that are more closely tied to the functional updates to the API documents, and
+reduces the inevitable "drift" between functional API and API specification.
+
+## User stories
+- As a dApp developer I want to have clear and accurate specification for all
+  Hedera APIs so that I can confidently design my application to call those
+  APIs.
+- As an independent implementor I want to have clear and accurate specification
+  for all Hedera APIs so that I can design an independent, correct, and
+  compatible implementation of Hedera network nodes, including a consensus node.
+- As an Hedera software contributor I want to have clear and accurate
+  specification for all Hedera APIs so that I can confidently implement changes
+  to Hedera-led software.
+
+## Specification
+This HIP makes sweeping changes to specification text for the existing API
+documents. These changes affect over 20,000 lines of text across almost 150
+files, which is excessive for a document such as this HIP. The full
+specification, therefore, is only linked here for brevity and clarity.<br/>
+The specification changes intended for this HIP are detailed in github as a
+[pull request](https://github.com/hashgraph/hedera-protobufs/pull/388)
+for the hedera-protobufs repository.
```

Comment:
Yes, for the actual improved specifications for each service/message/field in the protobufs - but I think it would be appropriate to put here in this HIP the _standards_ that have been used in those specifications in that PR which we intend to require all future protobuf additions/changes to use!

I'm not speaking of the standards we're going to use in _defining_ the protobufs themselves, (e.g., the section [_Content Guidelines_](https://github.com/hashgraph/hedera-protobufs/blob/02b1569675efa79db01c279d84b3295d23c0c810/documents/Specification-Format-Style-Guidelines.md?plain=1#L88) with, e.g., the preference of unsigned vs signed integer fields) but rather the stuff in [_Style Guidelines for "specification" comments_](https://github.com/hashgraph/hedera-protobufs/blob/02b1569675efa79db01c279d84b3295d23c0c810/documents/Specification-Format-Style-Guidelines.md?plain=1#L3) (where it says, e.g., "be prescriptive", all messages need a description but not all fields do, etc.).

* And the above with the associated examples given there in the PR.

In fact, I'll go so far as to suggest that the direction of this HIP be changed to _be_ the specification of how we'll both specify and document the specification of all protobufs in the future.  To contain _both_ of those sections I mentioned above.  And relegate the [PR itself](https://github.com/hashgraph/hedera-protobufs/pull/388) to not "be the HIP, just in a remote location" but instead to be the "reference implementation". That would seem to be the more appropriate use of the intent and process of HIPs (especially as the PR doesn't actually change anything in the protobufs themselves, only **greatly** improves the documentation).

And if that suggestion was taken then this HIP could also contain (eventually, not necessarily now) our _other_ requirements for protobufs - e.g., that we follow rules such that the encoding can be deterministic and that in fact the PBJ protobuf compiler must be used (if that's the case); this information is important for the user story of the "independent implementator" above and thus should be documented in this HIP.  (I don't think it is anywhere else.). Also how we _version_ protobufs for backward/forward compatibiility (I think we're just using standard protobuf "best practices" for that but even if so we could document them _here_ instead of relying on developers to just know them.)

Finally, mention should be made here in this "Specification" section that we're using protoc-gen-doc style comments.  It's mentioned above in the "Abstract" but the abstract is not the specification part of the HIP and in any event it is mentioned tentatively up there ("we should").

Reply:
This is a very good suggestion.
I've implemented the suggested changes as I understand them.  Please reply with further clarification if I missed part of your intent here.